### PR TITLE
[Xtensa] Fix FP mul-sub fusion (LLVM-276)

### PR DESCRIPTION
**Pull request by zRedShift**
`performMADD_MSUBCombine` treats `x * y ± z` and `z ± x * y` interchangeably which is wrong for `msub.s` which expects the latter. Added a check to determine that the orientation is correct, and if not, negate the result.

Seems like the original code was inspired by the MIPS code, which only has fused-multiply-add, and the order of the operands is irrelevant in that case.

This is my first PR to LLVM, so keep this in mind while reviewing, because I don't want to introduce a regression.

Related to (and hopefully fixes) esp-rs/rust#180, esp-rs/rust#185.

### CODE REVIEW

**Comment by zRedShift** [comment]
Correction: MIPS does in fact have `msub.(s|d)`, but the instruction shape is inverted to the way xtensa handles it
```
MSUB.S fd, fr, fs, ft
fd ← (fs × ft) - fr
```

Planning to add some llc/filechecker codegen tests for this fix

File: llvm/test/CodeGen/Xtensa/float-fma.ll [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+define float @fnmsub_s_swap_intrinsics(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fnmsub_s_swap_intrinsics:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    neg.s f10, f10
+; XTENSA-NEXT:    madd.s f10, f9, f8
+; XTENSA-NEXT:    rfr a2, f10
```
Comment (andreisfr) [inline llvm/test/CodeGen/Xtensa/float-fma.ll:142]: Isn't "fnmsub_s_swap_intrinsics" is the same as "fmsub_s_intrinsics" intrinsic?

File: llvm/test/CodeGen/Xtensa/float-fma.ll [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+; XTENSA-NEXT:    retw.n
+  %negc = fneg float %c
+  %fma = call float @llvm.fma.f32(float %a, float %b, float %negc)
```
Comment (andreisfr) [inline llvm/test/CodeGen/Xtensa/float-fma.ll:145]: @zRedShift , I think all is good and this PR is ready to merge. One question about "fnmsub_s_swap_intrinsics". It looks the same as "fmsub_s_intrinsics" intrinsic. Or I missed something?

Comment (zRedShift) [reply llvm/test/CodeGen/Xtensa/float-fma.ll:145]: @andreisfr just saw this in the email, and you're right. I copied these from LoongArch/float-fma.ll, and it looks like fnmsub_s_swap_intrinsics was supposed to be 
```llvm
  %negb = fneg float %b
  %fma = call float @llvm.fma.f32(float %a, float %negb, float %c)
  ret float %fma
```
I'll fix this

Comment (andreisfr) [reply llvm/test/CodeGen/Xtensa/float-fma.ll]: @zRedShift , I combined  the third commit with the second commit from this PR to single commit and merged all commits with main branch. Thank you very much!

**Comment by gerekon** [comment]
Closed with 19e97af664bf3581332f2e9c5d66b607a511c5c2

**Pull request closed without merging**